src/main/kotlin/com/github/vol0n/utbotcppclion/services/Client.kt

```diff
@@ -0,0 +1,276 @@
+                    "UTBot tried to configure project, but failed with the " +
+                            "following message: ${serverResponse.message}", project
+                )
+                else -> notifyUnknownResponse(serverResponse, project)
+            }
+        }
+
+        val request = getProjectConfigRequestMessage(project, Testgen.ConfigMode.GENERATE_JSON_FILES)
+        stub.configureProject(request).handleWithProgress(
+            "Ask server to generate json",
+            ::progressOrNull,
+            ::handleJSONGeneration,
+        )
+    }
+
+    private suspend fun heartBeat() {
+        try {
+            stub.heartbeat(Testgen.DummyRequest.newBuilder().build())
+            connectionChangedPublisher.onChange(connectionStatus, ConnectionStatus.CONNECTED)
+            connectionStatus = ConnectionStatus.CONNECTED
+        } catch (e: Exception) {
+            logger.info("Heartbeat failed: could not connect to server!")
+            connectionChangedPublisher.onChange(connectionStatus, ConnectionStatus.BROKEN)
+            connectionStatus = ConnectionStatus.BROKEN
+        }
+    }
+
+    private fun Flow<Testgen.TestsResponse>.handleWithProgress(progressName: String = "Generating Tests") {
+        this.handleWithProgress(
+            progressName, Testgen.TestsResponse::getProgress,
+            handleResponse = this@Client::handleTestResponse
+        )
+    }
+
+    // T: TestResponse | ProjectConfigResponse
+    fun <T> Flow<T>.handleWithProgress(
+        progressName: String,
+        progressAccessor: (T) -> Util.Progress?,
+        onCompleted: (T) -> Unit = {},
+        handleResponse: (T) -> Unit = {}
+    ) {
+        val uiProgress = UTBotRequestProgressIndicator(progressName)
+        // start showing progress in status bar
+        uiProgress.start()
+        uiProgress.requestJob = grpcCoroutineScope.launch {
+            withTimeout(STREAM_HANDLING_TIMEOUT) {
+                this@handleWithProgress
+                    .catch { exception -> logger.warn(exception.message) }
+                    .collect {
+                        val progress = progressAccessor(it)
+                        // when we receive last message from server stream
+                        if (progress == null || progress.completed) {
+                            onCompleted(it)
+                            return@collect
+                        }
+                        // update progress in status bar
+                        uiProgress.fraction = progress.percent
+                        uiProgress.text = progress.message
+                        handleResponse(it)
+                    }
+            }
+            uiProgress.complete()
+        }
+    }
+
+    companion object {
+        const val RANDOM_SEQUENCE_MAX_VALUE = 10
+        const val RANDOM_SEQUENCE_LENGTH = 5
+        const val HEARTBEAT_INTERVAL: Long = 500L
+        const val STREAM_HANDLING_TIMEOUT: Long = 5000L
```

Comment:
So the time limit for any test generation is 5 seconds? It is really not enough. I would refuse of such limitation. Test genration actually can run for hours on really big projects